gpflow/kernels.py

```diff
@@ -1005,6 +1006,94 @@ def Kdiag(self, X, presliced=False):
+
+    @incollection{ldgtg2014,
+      author = {Lloyd, James Robert and Duvenaud, David and Grosse, Roger and Tenenbaum, Joshua B. and Ghahramani, Zoubin},
+      title = {Automatic Construction and Natural-language Description of Nonparametric Regression Models},
+      booktitle = {Proceedings of the Twenty-Eighth AAAI Conference on Artificial Intelligence},
+      series = {AAAI'14},
+      year = {2014},
+      pages = {1242--1250},
+      numpages = {9},
+      url = {http://dl.acm.org/citation.cfm?id=2893873.2894066},
+      acmid = {2894066},
+      publisher = {AAAI Press},
+    }
+    """
+    def __init__(self, kernels, locations=0.0, widths=1.0, activation_order=None, name=None):
+        super().__init__(kernels, name=name)
+        if self.input_dim > 1:
+            raise ValueError("ChangePoints kernel only valid for 1d inputs")
+
+        if not isinstance(locations, Iterable):
+            locations = [locations]
+        if not isinstance(widths, Iterable):
+            widths = [widths]
+
+        self.locations = Parameter(locations, transform=None, dtype=settings.float_type)
+        self.widths = Parameter(widths, transform=transforms.positive, dtype=settings.float_type)
+        self.activation_order = list(activation_order or range(len(kernels)))
+
+        if not all(0 <= i < len(self.kernels) for i in self.activation_order):
+            raise ValueError("ChangePoints `activation_order` should be a list of integers "
+                             "specifying the order in which each kernel is active")
+
+        if len(self.activation_order) != len(locations) + 1:
+            raise ValueError("Number of active kernels ({a}) must be one more than the number of "
+                             "change-points ({cp})".format(a=len(self.activation_order), cp=len(locations)))
+
+        if len(widths) > 1 and len(widths) != len(locations):
+            raise ValueError("Number of widths ({nw}) does not match number of locations "
+                             "({nl})".format(nw=len(widths), nl=len(locations)))
+
+    @property
+    def num_changepoints(self):
+        """
+        Total number of change-points defined by kernel.
+        """
+        return len(self.activation_order) - 1
+
+    @params_as_tensors
+    def K(self, X, X2=None):
+        sig_X = self._sigmoids(X)
+        sig_X2 = self._sigmoids(X2) if X2 is not None else sig_X
+        starters = sig_X * tf.transpose(sig_X2, perm=(1, 0, 2))
+        stoppers = (1 - sig_X) * tf.transpose((1 - sig_X2), perm=(1, 0, 2))
+
+        regime_Ks = []
+        for i, ik in enumerate(self.activation_order):
+            K = self.kernels[ik].K(X, X2)
+            if i > 0:
+                K *= starters[:, :, i-1]
+            if i < self.num_changepoints:
+                K *= stoppers[:, :, i]
+            regime_Ks.append(K)
+
+        return reduce(tf.add, regime_Ks)
+
+    @params_as_tensors
+    def Kdiag(self, X):
+        return tf.matrix_diag_part(self.K(X))
```

Comment:
Could this be evaluated without constructing the full `K` matrix, by calling into the individual kernels' `Kdiag` instead?

Reply:
Yes absolutely - I was too lazy in this first iteration to figure out the right way to do that